README.md
## ContinuedFractions

* http://rubygems.org/gems/continued_fractions

#### Description

Generates quotients and convergents for a given number.

#### Synopsis

    require 'continued_fractions'
  
    pi = Math::PI
    => 3.14159265358979
    
    cf = ContinuedFraction.new(pi,10)
    => #<ContinuedFractions::ContinuedFraction:0x00000101101368 @number=3.14159265358979, @limit=10,
            @quotients=[3, 7, 15, 1, 292, 1, 1, 1, 2, 1],
            @convergents=[[0, 1], [1, 0], [3, 1], [22, 7], [333, 106],
                          [355, 113], [103993, 33102], [104348, 33215], [208341, 66317], [312689, 99532], [833719, 265381], [1146408, 364913]]>
                         
    cf.convergents
    => [[3,1], [22,7], [333,106], [355,113], [103993,33102], [104348,33215], [208341,66317], [312689,99532], [833719,265381], [1146408,364913]]
    
    # Convergents can be output as an array of Rationals too
    cf.convergents_as_rationals
    [(3/1), (22/7), (333/106), (355/113), (103993/33102), (104348/33215), (208341/66317), (312689/99532), (833719/265381), (1146408/364913)]
    
    cf.convergent(1)
    => [22,7]
    
    cf.quotients
    => [3, 7, 15, 1, 292, 1, 1, 1, 2, 1]
    
    # Add, subtract, multiply and divide continued fractions.  The limit of the result is the max limit of the operands.
    cf2 = ContinuedFraction.new(rand,20)
    => #<ContinuedFractions::ContinuedFraction:0x000001009dde50 @number=0.03834175394982653, @limit=20, @quotients=[0, 26, 12, 3, 4, 1, 2, 24, 5, 1, 1, 1, 2, 1, 2, 2, 1, 7, 1, 5], @convergents=[[0, 1], [1, 26], [12, 313], [37, 965], [160, 4173], [197, 5138], [554, 14449], [13493, 351914], [68019, 1774019], [81512, 2125933], [149531, 3899952], [231043, 6025885], [611617, 15951722], [842660, 21977607], [2296937, 59906936], [5436534, 141791479], [7733471, 201698415], [59570831, 1553680384], [67304302, 1755378799], [396092341, 10330574379]]>
    cf + cf2
    => #<ContinuedFractions::ContinuedFraction:0x000001009d2708 @number=3.1799344075396196, @limit=20, @quotients=[3, 5, 1, 1, 3, 1, 5, 3, 7, 25, 1, 2, 15, 6, 1, 1, 2, 3, 1, 2], @convergents=[[3, 1], [16, 5], [19, 6], [35, 11], [124, 39], [159, 50], [919, 289], [2916, 917], [21331, 6708], [536191, 168617], [557522, 175325], [1651235, 519267], [25326047, 7964330], [153607517, 48305247], [178933564, 56269577], [332541081, 104574824], [844015726, 265419225], [2864588259, 900832499], [3708603985, 1166251724], [10281796229, 3233335947]]>

#### Install

sudo gem install continued_fractions

#### Requirements/Dependencies

* Ruby 2.2+
* Depends on Ruby's Rational class

##### Developers

* RSpec required to run tests
* Echoe required to run rake tasks
  

CHANGELOG
v0.1.0 Initial release

v1.1.0 Change public class representation of convergents from an array of Rationals to an array of arrays.

v1.2.0 Change parameter signature of ContinuedFraction#convergents_as_rationals

v1.3.0 Add binary operators: + - * /

v1.4.0 Eliminate need to invoke "include ContinuedFractions"

v1.5.0 Remove ContinuedFractions module; Move its methods into ContinuedFraction class

v1.6.0 Handle rational numbers, finite continued fractions

spec/spec_helper.rb
$LOAD_PATH << File.join(File.dirname(__FILE__), "/../lib")
require 'rubygems'
require 'rspec'
require 'continued_fractions'

RSpec.configure do |config|
  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = "random"
end

spec/continued_fractions/continued_fraction_spec.rb
require File.join(File.dirname(__FILE__), "/../spec_helper")

describe ContinuedFraction do
  let(:cf) { described_class.new(number,10) }
  let(:number) { rand }
  
  it "returns an array of convergents" do
    expect(cf.convergents).to be_kind_of(Array)
  end
  
  it "preserves the number input" do
    expect(cf.number).to eq number
  end
  
  it "adds a number on the right-hand-side with a continued fraction and returns a continued fraction" do
    expect((cf + 3)).to be_kind_of(ContinuedFraction)
  end
  
  it "subtracts a number on the right-hand-side from a continued fraction and returns a continued fraction" do
    expect((cf - 3)).to be_kind_of(ContinuedFraction)
  end
  
  it "multiplies a number on the right-hand-side with a continued fraction and returns a continued fraction" do
    expect((cf * 3)).to be_kind_of(ContinuedFraction)
  end
  
  it "divides a number on the right-hand-side with a continued fraction and returns a continued fraction" do
    expect((cf / 3)).to be_kind_of(ContinuedFraction)
  end
  
  it "adds a continued fraction with another continued fraction and returns a continued fraction" do
    c = described_class.new(rand,20)
    expect((cf + c)).to be_kind_of(ContinuedFraction)
  end
  
  it "subtracts a continued fraction with another continued fraction and returns a continued fraction" do
    c = described_class.new(rand,20)
    expect((cf - c)).to be_kind_of(ContinuedFraction)
  end
  
  it "multiplies a continued fraction with another continued fraction and returns a continued fraction" do
    c = described_class.new(rand,20)
    expect((cf * c)).to be_kind_of(ContinuedFraction)
  end
  
  it "divides a continued fraction with another continued fraction and returns a continued fraction" do
    c = described_class.new(rand,20)
    expect((cf / c)).to be_kind_of(ContinuedFraction)
  end
  
  it "assigns the resulting continued fraction of a binary operation the max limit of the two operands" do
    c = described_class.new(rand,20)
    result = cf + c
    expect(result.limit).to eq [cf.limit,c.limit].max
  end
  
  context 'with irrational numbers' do
    it "accurately calculates the convergents" do
      # First 10 convergents of PI are...
      convs = ['3/1', '22/7', '333/106', '355/113', '103993/33102', '104348/33215', '208341/66317', '312689/99532', '833719/265381', '1146408/364913'].map(&:to_r)
      cf = described_class.new(Math::PI,10)
      expect((cf.convergents_as_rationals - convs)).to be_empty
    end
    
    it "contains convergents approaching the number" do
      0.upto(cf.convergents.length-2) do |i|
        convergent_rational_i_plus1, convergent_rational_i = cf.convergent_to_rational(cf.convergents[i+1]), cf.convergent_to_rational(cf.convergents[i])
        expect(((convergent_rational_i_plus1 - cf.number).abs <= (convergent_rational_i - cf.number).abs)).to be true
      end
    end
    
    it "contains convergents which are expressed in lowest terms" do
      1.upto(cf.convergents.length-1) do |i|
        convergent_rational_i, convergent_rational_i_plus1 = cf.convergent_to_rational(cf.convergent(i)), cf.convergent_to_rational(cf.convergent(i+1))
        expect((convergent_rational_i.numerator*convergent_rational_i_plus1.denominator - convergent_rational_i_plus1.numerator*convergent_rational_i.denominator)).to eq (-1)**(i)
      end
    end
  end
  
  context 'with rational numbers' do
    it 'changes the limit to the number of convergents calculated' do
      expect(described_class.new(1.5, 10).limit).to be 2
    end
    
    it 'calculates the convergents' do
      convs = [ '1/1', '3/2' ].map(&:to_r)
      expect(described_class.new(1.5, 10).convergents_as_rationals - convs).to be_empty
    end
  end
end

lib/continued_fractions.rb
# ContinuedFractions
#
# Generates quotients and convergents for a given number
#
# Author:: Jose Hales-Garcia (mailto:jolohaga@me.com)
# Copyright:: Copyright (c) 2015 Jose Hales-Garcia
#
#
class ContinuedFraction
  attr_accessor :number, :quotients, :limit
  
  # For a given number calculate its continued fraction quotients and convergents up to limit.
  #
  # The limit is 5 by default. Pass an integer in the second parameter to change it.
  #
  # Example:
  #   cf = ContinuedFraction.new(Math::PI,10)
  #   => #<ContinuedFractions::ContinuedFraction:0x000001010ed5c8 @number=3.14159265358979, @limit=10,
  #       @quotients=[3, 7, 15, 1, 292, 1, 1, 1, 2, 1],
  #       @convergents=[[0, 1], [1, 0], [3, 1], [22, 7], [333, 106],
  #                     [355, 113], [103993, 33102], [104348, 33215], [208341, 66317], [312689, 99532], [833719, 265381], [1146408, 364913]]>
  #
  def initialize(number,limit=5)
    @number = number
    @limit = limit
    @quotients = calculate_quotients
    @convergents = calculate_convergents
  end
  
  # Return nth convergent.
  #
  # Example:
  #   cf = ContinuedFraction.new(Math::PI,10)
  #   cf.convergent(3)
  #   => [355,113]
  #
  def convergent(nth)
    raise(IndexError, "Convergent index must be greater than zero.") unless nth > 0
    convergents[nth-1]
  end
  
  # Return array of convergents of the continued fraction instance up to the nth convergent as an array
  # comprising of numerators in [i,0] and denominators in [i,1].
  # If nth is nil, then return the entire list.
  #
  # Example:
  #   cf = ContinuedFraction.new(Math::PI,10)
  #   cf.convergents
  #   => [[3,1], [22,7], [333,106], [355,113], [103993,33102],
  #       [104348,33215], [208341,66317], [312689,99532], [833719,265381], [1146408,364913]]
  #
  # Or:
  #   cf.convergents(3)
  #   => [[3,1], [22,7], [333,106]]
  #
  def convergents(nth=nil)
    nth ||= @convergents.length
    @convergents[0...nth]
  end
  
  # Return array of convergents of the continued fraction instance converted as Rationals.
  # If nth is nil, then return the entire list.
  #
  # Example:
  #   cf = ContinuedFraction.new(Math::PI,10)
  #   cf.convergents_as_rationals
  #   => [(3/1), (22/7), (333/106), (355/113), (103993/33102),
  #       (104348/33215), (208341/66317), (312689/99532), (833719/265381), (1146408/364913)]
  #
  # Or:
  #   cf.convergents_as_rationals(3)
  #   => [(3/1), (22/7), (333/106)]
  #
  def convergents_as_rationals(nth=nil)
    nth ||= convergents.length
    convergents[0...nth].map{|convergent| convergent_to_rational(convergent)}
  end
  
  def +(other)
    number_of(other) do |num,prec|
      evaluate("#{number} + #{num}",prec)
    end
  end
  
  def -(other)
    number_of(other) do |num,prec|
      evaluate("#{number} - #{num}",prec)
    end
  end
  
  def /(other)
    number_of(other) do |num,prec|
      evaluate("#{number} / #{num}",prec)
    end
  end
  
  def *(other)
    number_of(other) do |num,prec|
      evaluate("#{number} * #{num}",prec)
    end
  end
  
  def convergent_to_rational(convergent) #:nodoc:
    Rational(convergent[0],convergent[1])
  end
  
  private
  def evaluate(exp, prec) #:nodoc:
    ContinuedFraction.new(eval(exp), prec)
  end
  
  def number_of(n) #:nodoc:
    num = nil
    prec = nil
    case n
    when Fixnum, Integer
      num = n
      prec = limit
    when ContinuedFraction
      num = n.number
      prec = [n.limit,limit].max
    end
    yield(num,prec)
  end

  def calculate_quotients #:nodoc:
    n = number
    Array.new(limit).tap do |qs|
      limit.times do |i|
        qs[i] = n.to_i
        break if ((divisor = n-qs[i]) == 0.0)
        n = 1.0/divisor
      end
      self.limit = qs.compact.length
    end.compact
  end

  def calculate_convergents #:nodoc:
    convs = convergence_matrix(quotients.length+2,2,1)
    nth ||= convs.length
    2.upto(quotients.length+1) do |i|
      i_minus1,i_minus2 = i-1,i-2
      convs[i][0] = convs[i_minus1][0]*quotients[i_minus2]+convs[i_minus2][0]
      convs[i][1] = convs[i_minus1][1]*quotients[i_minus2]+convs[i_minus2][1]
    end
    convs[2...nth+2]
  end
  
  def convergence_matrix(n,m,fill=nil) #:nodoc:
    conv_mat = Array.new(n).map!{Array.new(m,fill)}
    conv_mat[0][0],conv_mat[1][1] = 0,0
    conv_mat
  end
end